File: JavaScript/src/player.js
export default class Player {
  /** This player's board value. */
  value;

  /**
   * Creates a player with `value`.
   * @param {number} value This player's numerical value.
   */
  constructor(value) {
    this.value = value;
  }

  static Empty = new Player(0);
  static X = new Player(1);
  static O = new Player(2);

  /**
   * Gets the next player to move after this player.
   * @returns {Player} The next player.
   */
  nextPlayer() {
    return this.value === Player.X.value ? Player.O : Player.X;
  }

  /**
   * Gets the icon to use for this player in a terminal window.
   * @param {boolean} colorize Whether or not to colorize the icon.
   * @returns {string} This player's icon.
   */
  icon(colorize = true) {
    // Get the base icon to use.
    var icon;
    switch (this.value) {
      case Player.X.value:
        icon = 'X';
        break;
      case Player.O.value:
        icon = 'O';
        break;
      default:
        icon = ' ';
    }

    // Colorize the icon if possible.
    if (colorize && this.value != Player.Empty.value) {
      var code;
      switch (this.value) {
        case Player.X.value:
          code = cyanEscapeCode;
          break;
        case Player.O.value:
          code = magentaEscapeCode;
          break;
      }

      // Colorize using the proper code.
      if (code) {
        icon = [code, icon, resetEscapeCode].join('');
      }
    }

    return icon;
  }
}

// Escape codes for terminal colorization.
const cyanEscapeCode = '\u001b[36m';
const magentaEscapeCode = '\u001b[35m';
const resetEscapeCode = '\u001b[0m';

File: Swift/Sources/CoreGame/Player.swift
/// A Tic-Tac-Toe player represented by their icon.
/// 
/// This player's `rawValue` will be represented by a `String`.
public enum Player: String {
    /// An empty spot.
    case empty = " "
    /// A spot taken by the "X" player.
    case x = "X"
    /// A spot taken by the "O" player.
    case o = "O"

    /// The player that should play next after this player.
    public var nextPlayer: Player {
        // You can use a shorthand notation for enums when the type is known.
        // In this case, Swift knows a `Player` should be returned, so you can
        // type `.x` or `.o` instead of `Player.x` and `Player.o`.
        // 
        // Also, no exlicit `return` is necessary for one-liners.
        self == .x ? .o : .x
    }

    /// Gets an icon to use for this `Player`. 
    /// 
    /// Including `colorize` will add escape codes for a colorful output when 
    /// printed on the command line.
    /// - Parameter colorize: Whether the icon should include terminal colors.
    /// - Returns: The icon to display for this player.
    public func icon(colorize: Bool = true) -> String {
        guard colorize else { return rawValue }
        let cyanEscapeCode = "\u{001b}[36m";
        let magentaEscapeCode = "\u{001b}[35m";
        let resetEscapeCode = "\u{001b}[0m";

        // Swift switches must be exhaustive or include a `default` statement.
        switch self {
            case .empty: return rawValue
            case .x: return cyanEscapeCode + rawValue + resetEscapeCode
            case .o: return magentaEscapeCode + rawValue + resetEscapeCode
        }
    }

}
